sdk/messaging/azservicebus/amqp_message.go

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azservicebus
+
+import (
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal/go-amqp"
+)
+
+// AMQPMessage represents the AMQP message, as received from Service Bus.
+// For details about these properties, refer to the AMQP specification:
+//   https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#section-message-format
+//
+// Some fields in this struct are typed 'any', which means they will accept AMQP primitives, or in some
+// cases slices and maps.
+//
+// AMQP simple types include:
+// - int (any size), uint (any size)
+// - float (any size)
+// - string
+// - bool
+// - time.Time
+type AMQPMessage struct {
+	// ApplicationProperties corresponds to the "application-properties" section of an AMQP message.
+	//
+	// The values of the map are restricted to AMQP simple types, as listed in the comment for AMQPMessage.
+	ApplicationProperties map[string]any
+
+	// Body represents the body of an AMQP message.
+	Body AMQPMessageBody
+
+	// DeliveryAnnotations corresponds to the "delivery-annotations" section in an AMQP message.
+	//
+	// The values of the map are restricted to AMQP simple types, as listed in the comment for AMQPMessage.
+	DeliveryAnnotations map[any]any
+
+	// DeliveryTag corresponds to the delivery-tag property of the TRANSFER frame
+	// for this message.
+	DeliveryTag []byte
+
+	// Footer is the transport footers for this AMQP message.
+	//
+	// The values of the map are restricted to AMQP simple types, as listed in the comment for AMQPMessage.
+	Footer map[any]any
+
+	// Header is the transport headers for this AMQP message.
+	Header *AMQPMessageHeader
+
+	// MessageAnnotations corresponds to the message-annotations section of an AMQP message.
+	//
+	// The values of the map are restricted to AMQP simple types, as listed in the comment for AMQPMessage.
+	MessageAnnotations map[any]any
+
+	// Properties corresponds to the properties section of an AMQP message.
+	Properties *AMQPMessageProperties
+
+	linkName string
```

Comment:
While not unique to this change, having unexported fields on a POD type means customers can't create a fake `AMQPMessage` for testing purposes.

Reply:
I suspect there's a fair amount of work that needs to be done in order to support mocks.  Also, I don't know what our actual testing strategy and recommendation is for `azservicebus`; this was just a general observation.  Feel free to open a tracking issue to address it later.

Reply:
@jhendrixMSFT: Just to check I'm getting what you mean - customers can create azservicebus.AMQPMessage{}, they just won't be able to to initialize those unexported fields. That's actually okay. They're only used when you actually call an operation (like Receiver.CompleteMessage) where they would have to be filled in.

For just mocking (ie, I'm passing an azservicebus.AMQPMessage around) it should be fine.

Reply:
Good point, I think I overgeneralized a bit (whenever I see exported POD types that mix un/exported fields my inner alarm goes off).  Agreed that passing around a mock version of this to/from mock APIs would be fine.  And it looks like there aren't any public APIs on this type that depend on the unexported fields so it should be fine.